src/engraving/rendering/dev/chordlayout.cpp

```diff
@@ -2224,34 +2250,37 @@ void ChordLayout::placeDots(const std::vector<Chord*>& chords, const std::vector
 void ChordLayout::layoutChords3(const MStyle& style, const std::vector<Chord*>& chords,
                                 const std::vector<Note*>& notes, const Staff* staff, LayoutContext& ctx)
 {
-    std::vector<Note*> leftNotes;   // notes to left of origin
-    leftNotes.reserve(8);
-
     Fraction tick      =  notes.front()->chord()->segment()->tick();
     double sp           = staff->spatium(tick);
     double stepDistance = sp * staff->lineDistance(tick) * .5;
     int stepOffset     = staff->staffType(tick)->stepOffset();
 
     double upDotPosX    = 0.0;
     double downDotPosX  = 0.0;
+    std::array<double, 12> dotPos{};
 
     int nNotes = int(notes.size());
 
     for (int i = nNotes - 1; i >= 0; --i) {
         Note* note     = notes[i];
         DirectionV dotPosition = note->userDotPosition();
-        if (note->chord()->dots()) {
+        const Chord* chord = note->chord();
+        if (chord->dots()) {
             if (dotPosition == DirectionV::AUTO && nNotes > 1 && note->visible() && !note->dotsHidden()) {
                 // resolve dot conflicts
                 int line = note->line();
-                Note* above = (i < nNotes - 1) ? notes[i + 1] : 0;
-                if (above && (!above->visible() || above->dotsHidden() || above->chord()->dots() == 0)) {
-                    above = 0;
+                Note* above = (i < nNotes - 1) ? notes[i + 1] : nullptr;
+                if (above
+                    && (!above->visible() || above->dotsHidden() || above->chord()->dots() == 0 || !chord->combineVoice()
+                        || !above->chord()->combineVoice())) {
```

Comment:
I see lots of instances where you're evaluating conditions like `chord1->combineVoice() && chord2->combineVoice()` or its negation. You could probably benefit from introducing a method in the Chord class for this. It could be something like 
`bool combineVoice(const Chord* otherChord) const { return combineVoice() && otherChord->combineVoice(); }`. 
Or actually even better as a static method:
`static bool combineVoice(const Chord* chord1, const Chord* chord2) { return chord1->combineVoice() && chord2->combineVoice(); }`
and maybe tag it as `inline` to make it zero-cost. 
That would make these things a bit more readable.

Reply:
There are 2 `&&` and 2 `||` comparisons so far - I'm tempted to leave them as is. I think static `combineVoiceAnd` and `combineVoiceOr` functions would be overkill

Reply:
You don't need two: the || comparison here is just the negation of the && comparison. In boolean logic `!(a && b) == !a || !b`. So if you define 
`static bool Chord::combineVoices(Chord* chord1, Chord* chord2) { return chord1->combineVoice() && chord2->combineVoice(); }`, 
then above you will call `combineVoices(chord1, chord2)` and here you will call `!combineVoices(chord1, chord2)`, which makes it clearer what this means and easier to read :)